clients/feeder/timeouts_utils.go

```diff
@@ -0,0 +1,128 @@
+package feeder
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+)
+
+const (
+	maxTimeout        = 2 * time.Minute
+	defaultTimeout    = 5 * time.Second
+	growthFactorFast  = 1.5
+	growthFactorSlow  = 1.2
+	fastGrowThreshold = 45 * time.Second
+)
+
+type TimeoutsList []time.Duration
+
+func generateTimeoutsListFromInitial(initial time.Duration, count int) TimeoutsList {
+	timeouts := make(TimeoutsList, count)
+	if count == 0 {
+		return timeouts
+	}
+	timeouts[0] = initial
+
+	for i := 1; i < count; i++ {
+		prev := timeouts[i-1]
+		next := getNextTimeout(prev)
+		timeouts[i] = next
+	}
+
+	return timeouts
+}
+
+func getTimeouts(timeouts TimeoutsList, maxRetries int) TimeoutsList {
+	if len(timeouts) == 0 {
+		return generateTimeoutsListFromInitial(defaultTimeout, maxRetries+1)
+	} else {
+		if len(timeouts) > 0 {
+			sort.Slice(timeouts, func(i, j int) bool {
+				return timeouts[i] < timeouts[j]
+			})
+		}
+
+		if len(timeouts) > maxRetries+1 {
+			return timeouts[:maxRetries+1]
+		} else {
+			count := maxRetries + 1 - len(timeouts)
+			next := getNextTimeout(timeouts[len(timeouts)-1])
+			timeouts := append(timeouts, generateTimeoutsListFromInitial(next, count)...)
+			return timeouts
+		}
+	}
+}
+
+func getNextTimeout(prev time.Duration) time.Duration {
+	var next time.Duration
+	if prev < fastGrowThreshold {
+		seconds := math.Ceil(float64(prev.Seconds()) * growthFactorFast)
+		return time.Duration(seconds) * time.Second
+	} else {
+		seconds := math.Ceil(float64(prev.Seconds()) * growthFactorSlow)
+		next = time.Duration(seconds) * time.Second
+	}
+	if next > maxTimeout {
+		next = maxTimeout
+	}
+	return next
+}
+
+func ParseTimeouts(values []string) (TimeoutsList, error) {
+	if len(values) == 0 {
+		return TimeoutsList{defaultTimeout}, nil
+	}
+
+	timeouts := make(TimeoutsList, 0, len(values))
+	for _, v := range values {
+		d, err := time.ParseDuration(v)
+		if err != nil {
+			return nil, fmt.Errorf("parsing timeouts: %v", err)
+		}
+		if d > maxTimeout {
+			d = maxTimeout
+		}
```

Comment:
Why do you have this constraint, If people want to have a timeout of 10000 hours, it's their decision

Reply:
Also, you have to have a check that number `i` is always bigger than `i-1`. Please add a test for this as well.

Reply:
Yes agree with the above. Right now the values are being sorted in `getTimeouts` and tested in `TestGetTimeouts`